pkg/certmaker/certmaker.go

```diff
@@ -0,0 +1,464 @@
+	var sv signature.SignerVerifier
+	var err error
+
+	switch config.Type {
+	case "awskms":
+		ref := fmt.Sprintf("awskms:///%s", keyID)
+		if config.Region != "" {
+			os.Setenv("AWS_REGION", config.Region)
+		}
+		sv, err = kms.Get(ctx, ref, crypto.SHA256)
+		if err != nil {
+			return nil, fmt.Errorf("failed to initialize AWS KMS: %w", err)
+		}
+
+	case "gcpkms":
+		ref := fmt.Sprintf("gcpkms://%s", keyID)
+		sv, err = kms.Get(ctx, ref, crypto.SHA256)
+		if err != nil {
+			return nil, fmt.Errorf("failed to initialize GCP KMS: %w", err)
+		}
+
+	case "azurekms":
+		keyURI := keyID
+		if strings.HasPrefix(keyID, "azurekms:name=") {
+			nameStart := strings.Index(keyID, "name=") + 5
+			vaultIndex := strings.Index(keyID, ";vault=")
+			if vaultIndex != -1 {
+				keyName := strings.TrimSpace(keyID[nameStart:vaultIndex])
+				vaultName := strings.TrimSpace(keyID[vaultIndex+7:])
+				keyURI = fmt.Sprintf("azurekms://%s.vault.azure.net/%s", vaultName, keyName)
+			}
+		}
+		if config.Options != nil && config.Options["tenant-id"] != "" {
+			os.Setenv("AZURE_TENANT_ID", config.Options["tenant-id"])
+			os.Setenv("AZURE_ADDITIONALLY_ALLOWED_TENANTS", "*")
+		}
+		os.Setenv("AZURE_AUTHORITY_HOST", "https://login.microsoftonline.com/")
+
+		sv, err = kms.Get(ctx, keyURI, crypto.SHA256)
+		if err != nil {
+			return nil, fmt.Errorf("failed to initialize Azure KMS: %w", err)
+		}
+
+	case "hashivault":
+		keyURI := fmt.Sprintf("hashivault://%s", keyID)
+		if config.Options != nil {
+			if token := config.Options["token"]; token != "" {
+				os.Setenv("VAULT_TOKEN", token)
+			}
+			if addr := config.Options["address"]; addr != "" {
+				os.Setenv("VAULT_ADDR", addr)
+			}
+		}
+
+		sv, err = kms.Get(ctx, keyURI, crypto.SHA256)
+		if err != nil {
+			return nil, fmt.Errorf("failed to initialize HashiVault KMS: %w", err)
+		}
+
+	default:
+		return nil, fmt.Errorf("unsupported KMS type: %s", config.Type)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to get KMS signer: %w", err)
+	}
+	if sv == nil {
+		return nil, fmt.Errorf("KMS returned nil signer")
+	}
+
+	return sv, nil
+}
+
+// CreateCertificates creates certificates using the provided KMS and templates.
+// It creates 3 certificates (root -> intermediate -> leaf) if intermediateKeyID is provided,
+// otherwise creates just 2 certs (root -> leaf).
+func CreateCertificates(sv signature.SignerVerifier, config KMSConfig,
+	rootTemplatePath, leafTemplatePath string,
+	rootCertPath, leafCertPath string,
+	intermediateKeyID, intermediateTemplatePath, intermediateCertPath string) error {
+
+	// Create root cert
+	rootTmpl, err := ParseTemplate(rootTemplatePath, nil)
+	if err != nil {
+		return fmt.Errorf("error parsing root template: %w", err)
+	}
+
+	// Get public key from signer
+	rootPubKey, err := sv.PublicKey()
+	if err != nil {
+		return fmt.Errorf("error getting root public key: %w", err)
+	}
+
+	rootCert, err := x509util.CreateCertificate(rootTmpl, rootTmpl, rootPubKey, signerWrapper{sv})
```

Comment:
Why not use `x509.CreateCertificate` from the standard library?

Reply:
this is so we can use stepca templates via "go.step.sm/crypto/x509util"